packages/flutter/test/widgets/editable_text_test.dart

```diff
@@ -478,6 +478,43 @@ void main() {
         equals('TextInputAction.done'));
   });
 
+  testWidgets('can only show toolbar when there is text and a selection',
+      (WidgetTester tester) async {
+    await tester.pumpWidget(
+      MaterialApp(
+        home: EditableText(
+          backgroundCursorColor: Colors.grey,
+          controller: controller,
+          focusNode: focusNode,
+          style: textStyle,
+          cursorColor: cursorColor,
+          selectionControls: materialTextSelectionControls,
+        ),
+      ),
+    );
+
+    final EditableTextState state =
+        tester.state<EditableTextState>(find.byType(EditableText));
+
+    expect(state.showToolbar(), false);
+    await tester.pump();
+    expect(find.text('PASTE'), findsNothing);
+
+    controller.text = 'blah';
+    expect(state.showToolbar(), false);
+    await tester.pump();
+    expect(find.text('PASTE'), findsNothing);
+
+    // Select something. Doesn't really matter what.
+    state.renderEditable.selectWordsInRange(
+      from: const Offset(0, 0),
+      cause: SelectionChangedCause.tap,
```

Comment:
I'm not deprecating SelectionChangedCause (my branch name is a misnomer I guess :D). We're still trying to be honest™️ with the real cause to supply it to users of the existing APIs via the various onSelectionChanged and whatnot. We're just not using it internally anymore, but we're not breaking external users. 

In this test, I'm intending for it to mean a single tap. 

I'm not making the select*'s cause optional because they still aren't (to keep supporting onSelectionChanged users who might care). 

So far, there are no 'fake' selection causes yet to have a new enum value.

Reply:
> So far, there are no 'fake' selection causes yet to have a new enum value.

What programmatic changes to the selection that are not triggered by an input event?

Reply:
Programmatic changes actually circumvents the whole thing it appears. i.e. nothing is triggered through the RenderEditable directly, no on*Changed is bubbled back through EditableText, there are no SelectionChangedCause involved, no toolbars are ever triggered and no on*Changed is called. 

This is in the various `@template flutter.widgets.editableText.onChanged` docs. `This callback does run not when the TextField's text is changed, via the TextField's [controller].`

Adding some additional docs to RenderEditable.select* to indicate that it's used for translating UI manipulations to logical selections and if you own the TextEditingController, it's a lot easier to go manipulate that directly.

Reply:
Thanks for clarifying that. So the rough summary would be: the RenderEditable selection changing methods are only called in response to user input.  That makes sense.

Reply:
It's a good question though. I only thought about it after you asked. But the new docs should address it now.